src/acl.c

```diff
@@ -550,22 +560,58 @@ void ACLSetSelectorCommandBitsForCategoryLogic(dict *commands, aclSelector *sele
             ACLChangeSelectorPerm(selector,cmd,value);
         }
         if (cmd->subcommands_dict) {
-            ACLSetSelectorCommandBitsForCategoryLogic(cmd->subcommands_dict, selector, cflag, value);
+            ACLSetSelectorCommandBitsForCategory(cmd->subcommands_dict, selector, cflag, value);
         }
     }
     dictReleaseIterator(di);
 }
 
-/* This is like ACLSetSelectorCommandBit(), but instead of setting the specified
- * ID, it will check all the commands in the category specified as argument,
- * and will set all the bits corresponding to such commands to the specified
- * value. Since the category passed by the user may be non existing, the
- * function returns C_ERR if the category was not found, or C_OK if it was
- * found and the operation was performed. */
-int ACLSetSelectorCommandBitsForCategory(aclSelector *selector, const char *category, int value) {
+/* Set the ACL category on the selector. Which categories are added
+ * are maintained so they can be added verbatim when serializing
+ * the user. */
+int ACLSetSelectorCategory(aclSelector *selector, const char *category, int value) {
     uint64_t cflag = ACLGetCommandCategoryFlagByName(category);
     if (!cflag) return C_ERR;
-    ACLSetSelectorCommandBitsForCategoryLogic(server.orig_commands, selector, cflag, value);
+    /* Add the new category to the list of categories */
+    sds new_category = sdsnew(category);
+    sdstolower(new_category);
+
+    /* Remove the category from the list if it was previously added */
+    char *existing = strstr(selector->categories, new_category);
+    int need_space = 1;
+    if (existing) {
+        char *copy_ptr = existing + sdslen(new_category);
+        existing -= 2;
+        if (*copy_ptr != '\0') {
+            /* If we aren't at the end of the categories,
+             * also remove the trailing space. */
+            copy_ptr++;
+        } else if (existing != selector->categories) {
+            /* Keep track of if we have a leading space. */
+            need_space = 0;
+        }
+
+        while (*copy_ptr) {
+            *existing = *copy_ptr;
+            existing++;
+            copy_ptr++;
+        }
+
+        /* Update the length here, but don't resize the allocation. We
+         * will re-use the space next when we append the category to
+         * the end. */
+        *existing = '\0';
+        sdsupdatelen(selector->categories);
+    }
```

Comment:
This is a little less code than maintaining a list, since we don't actually need to go search through each list element to find duplicates and we don't need to concat the stuff together when displaying. I'm fine switching if you feel strongly about the simpler code.